homeassistant/components/binary_sensor/generic_hygrostat.py

```diff
@@ -0,0 +1,217 @@
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.generic_hygrostat/
+"""
+import asyncio
+import collections
+import time
+from datetime import timedelta
+import logging
+
+import voluptuous as vol
+
+from homeassistant.core import callback
+from homeassistant.components.climate import PLATFORM_SCHEMA
+from homeassistant.const import (STATE_ON, STATE_OFF, STATE_UNKNOWN, CONF_NAME)
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['sensor']
+
+SAMPLE_INTERVAL = timedelta(minutes=5)
+
+DEFAULT_NAME = 'Generic Hygrostat'
+
+ATTR_SENSOR_ID = 'sensor'
+ATTR_DELTA_TRIGGER = 'delta_trigger'
+ATTR_TARGET_OFFSET = 'target_offset'
+ATTR_MAX_ON_TIME = 'max_on_time'
+ATTR_SENSOR_HUMIDITY = 'sensor_humidity'
+ATTR_TARGET = 'target'
+ATTR_SAMPLES = 'samples'
+ATTR_MAX_ON_TIMER = 'max_on_timer'
+
+CONF_SENSOR = 'sensor'
+CONF_DELTA_TRIGGER = 'delta_trigger'
+CONF_TARGET_OFFSET = 'target_offset'
+CONF_MAX_ON_TIME = 'max_on_time'
+
+DEFAULT_DELTA_TRIGGER = 3
+DEFAULT_TARGET_OFFSET = 3
+DEFAULT_MAX_ON_TIME = 7200
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SENSOR): cv.entity_id,
+    vol.Optional(CONF_DELTA_TRIGGER, default=DEFAULT_DELTA_TRIGGER):
+        vol.Coerce(float),
+    vol.Optional(CONF_TARGET_OFFSET, default=DEFAULT_TARGET_OFFSET):
+        vol.Coerce(float),
+    vol.Optional(CONF_MAX_ON_TIME, default=DEFAULT_MAX_ON_TIME):
+        vol.Coerce(float)
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Generic Hygrostat platform."""
+    name = config.get(CONF_NAME)
+    sensor_id = config.get(CONF_SENSOR)
+    delta_trigger = config.get(CONF_DELTA_TRIGGER)
+    target_offset = config.get(CONF_TARGET_OFFSET)
+    max_on_time = config.get(CONF_MAX_ON_TIME)
+
+    async_add_devices([GenericHygrostat(
+        hass, name, sensor_id, delta_trigger, target_offset, max_on_time)])
+
+
+class GenericHygrostat(Entity):
+    """Representation of a Generic Hygrostat device."""
+
+    def __init__(self, hass, name, sensor_id, delta_trigger, target_offset,
+                 max_on_time):
+        """Initialize the hygrostat."""
+        self.hass = hass
+        self._name = name
+        self.sensor_id = sensor_id
+        self.delta_trigger = delta_trigger
+        self.target_offset = target_offset
+        self.max_on_time = max_on_time
+
+        self.sensor_humidity = None
+        self.target = None
+        self.samples = collections.deque([], 3)
+        self.max_on_timer = None
+
+        self._state = STATE_OFF
+        self._icon = 'mdi:water-percent'
+
+        self._async_update()
+
+        async_track_time_interval(hass, self._async_update, SAMPLE_INTERVAL)
```

Comment:
Why don't you subscribe to the humidity sensor state updates and use async_call_later with the delay = max_on_time ?

Reply:
First of all, the max_on_time is just a safety mechanism, it has nothing todo with the humidity sampling routine. So I think you mean delay = SAMPLE_INTERVAL?

The current humidity sensor I use is updating very irregular, sometimes within a few seconds, sometimes it can take up 20 minutes. I need a humidity sample every 5 minutes!

Maybe I'm getting you wrong?

Reply:
What I'm saying is that the all component can be implemented without pooling (sampling).

There is no point in sampling the humidity sensor if he hasn't changed. If your humidity sensor is not working properly (sampling wise), that is something that must be correct in that sensor code. Not in this component.

Reply:
No! This is not how moving averages work, you need to calculate them over time, not when the sensor updates. A sensor that updates every 5 minutes and rises 5% is telling you a hole different story than a sensor that updates only when a change by x% in humidity occurs.

Just like the sma and ema indicators work with trading, they are used over time, not when there are buy/sell orders ;-)

Reply:
When you get an event you get the time the event happened. 

You just need to do the math taking into consideration how long each humidity value remained at a given value. 

Doing this this away will mean you have less error too, since your sample might be biased (eg. you might be loosing important variations between samples)

Reply:
Paper looks interesting. Don't know if I can fix it in the short term, do not have much time right now... 

How do I fix the max_on_timer? With the async_track_point_in_utc_time helper?

Reply:
I would not call it a fix... but yes you can use async_track_point_in_utc_time to set a time in the future = max_on_timer to turn off the fan.